blog/content/post/2025-04-18-cache-attack-mitigation-testing.md

```diff
@@ -0,0 +1,350 @@
+  used by the first program (which works in background)
+
+  ```c
+  for (void *addr = range.start; addr < range.end; addr += CACHE_LINE_SIZE) {
+      libflush_flush(libflush_session, addr);
+      libflush_memory_barrier();
+      usleep(RELOAD_WAIT_US);
+      libflush_memory_barrier();
+      uint64_t time = libflush_reload_address(libflush_session, addr);
+      if (time <= threshold) // cache hit
+          cache_line_hit[(addr - range.start) / CACHE_LINE_SIZE] += 1;
+  }
+  ```
+
+Below are results which I achieved. First program was accessing indices 0, 8,
+and 32:
+
+```text
+Addr          Cache hits      Cache line offset       Byte offset     uint64_t offset
+0x7fb626522000  1               +0                      +0              +<0-8>
+0x7fb626522380  1               +14                     +896            +<112-120>
+0x7fb626522580  1               +22                     +1408           +<176-184>
+0x7fb626522600  1               +24                     +1536           +<192-200>
+0x7fb6265226c0  1               +27                     +1728           +<216-224>
+0x7fb6265222c0  15              +11                     +704            +<88-96>
+0x7fb626522340  27              +13                     +832            +<104-112>
+0x7fb626522240  31              +9                      +576            +<72-80>
+0x7fb626522280  32              +10                     +640            +<80-88>
+```
+
+Since we are using shared memory, and we know what it contains, it's very easy
+to identify the indices that the second program accessed. The start of the array
+is at cache line with offset +9. From the collected data, we can easily extract
+index ranges: `<0–8)`, `<8–16)`, `<32–40)`. In this case, the test also detected
+access to the indices in `<16–24)` range (about half as often), and singular
+accesses to couple other which we did not use.
+
+The exact cause is unknown, but I suspect it's due to architectural
+optimizations (the processor noticed that we accessed address `+0x40` and
+`+0x80`, and assumed we might also want to access address `+0xc0`).
+
+In the case of `Evict+Time`, I managed to get satisfactory results, but only
+when I used the shared function directly. In such a case, the difference in
+execution time was `<-7, 28>` units when evicting unused addresses, and around
+`<300, 400>` units when evicting addresses used by the function.
+
+```c
+size_t indices[] = {0};
+libflush_flush(libflush_session, addr);
+libflush_memory_barrier();
+uint64_t time = libflush_get_timing(libflush_session);
+libflush_memory_barrier();
+access_array(1, indices);
+libflush_memory_barrier();
+time = libflush_get_timing(libflush_session) - time;
+```
+
+I am still working on `Prime+Probe` example.
+
+## Target testing platform
+
+The main goal of these tests will be verifying whether the cache coloring
+implementation in the [CROSSCON](https://github.com/crosscon) hypervisor
+prevents inter-VM cache attacks.
```

Comment:
Cool, I need that in the introduction and abstract. The goal is to give a sneak peek trailer of what we are trying to achieve and what can be expected in the article.